ws_server.go

```diff
@@ -1,48 +1,117 @@
 package introspection
 
 import (
-	"context"
+	"errors"
+	"fmt"
+	"net"
 	"net/http"
+	"sync"
 	"time"
 
+	"github.com/libp2p/go-libp2p-core/introspection"
+
+	logging "github.com/ipfs/go-log"
+
 	"github.com/gogo/protobuf/proto"
 	"github.com/gorilla/websocket"
-	logging "github.com/ipfs/go-log"
-	"github.com/libp2p/go-libp2p-core/introspect"
 )
 
 var logger = logging.Logger("introspection-server")
 var upgrader = websocket.Upgrader{}
 
-// StartServer starts the ws introspection server with the given introspector
-func StartServer(introspector introspect.Introspector) func() error {
-	// introspect handler
-	http.HandleFunc("/introspect", wsUpgrader(introspector))
+type WsServer struct {
+	sync.RWMutex
+
+	config    *WsServerConfig
+	listeners []net.Listener
+	server    *http.Server
+	closeWg   sync.WaitGroup
+}
+
+type WsServerConfig struct {
+	ListenAddrs []string
+}
+
+// NewWsServer creates a WebSockets server to serve introspection data.
+func NewWsServer(introspector introspection.Introspector, config *WsServerConfig) (*WsServer, error) {
+	mux := http.NewServeMux()
+	// introspection handler
+	mux.HandleFunc("/introspect", wsUpgrader(introspector))
+
+	srv := &WsServer{
+		server: &http.Server{Handler: mux},
+		config: config,
+	}
+	return srv, nil
+}
+
+// Start starts this WS server.
+func (s *WsServer) Start() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.listeners) > 0 {
+		return errors.New("failed to start WS server: already started")
+	}
 
-	// start server
-	srv := http.Server{
-		// TODO Need a better strategy to select an address
-		Addr: introspector.ListenAddrs()[0],
+	if len(s.config.ListenAddrs) == 0 {
```

Comment:
@raulk A nil config will panic here as we don't check for this in `NewWsServer `.